Collecting tests from Python modules not generated by Python files

I'm a core developer of [the Hy programming language](https://github.com/hylang/hy), which is a Lisp syntax for Python, and I'm trying to migrate its test suite from Nose to pytest. The repository has a lot of tests written in Python but also tests written in Hy. Hy itself is implemented in Python and provides [a meta-importer](https://github.com/hylang/hy/blob/master/hy/importer.py) so that a Python program can import a Hy module just like a Python module. So in theory, pytest should be able to collect tests from Hy files just as well as Python files, so long as Hy's meta-importer is active. Once a Hy file is loaded, there's basically no difference between Hy and Python; a Hy module is in fact a Python module.

I found a hacky way to run the Hy tests, using [this conftest.py](https://github.com/Kodiologist/hy/blob/pytest/conftest.py), which calls `_pytest.python.pytest_pycollect_makemodule` in a `pytest_collect_file` hook. The catch is that this hits `assert name.endswith(".py")` in `_pytest.python`, and Hy files end with `.hy`, not `.py`. If the assertion is commented out, everything works.

Is there a better way to do this? If not, perhaps you should remove the assertion.

I'm using pytest 3.0.7. Hy is intended to support, and is tested on, Python 2.7 and Python 3.3 and up.

Sorry, here's an example of the backtraces I get when running pytest on this repository without the module-name spoof:

    _______________ ERROR collecting tests/native_tests/extra/reserved.hy ________________
    ../lib/python3.6/site-packages/_pytest/runner.py:163: in __init__
        self.result = func()
    ../lib/python3.6/site-packages/_pytest/main.py:476: in _memocollect
        return self._memoizedcall('_collected', lambda: list(self.collect()))
    ../lib/python3.6/site-packages/_pytest/main.py:347: in _memoizedcall
        res = function()
    ../lib/python3.6/site-packages/_pytest/main.py:476: in <lambda>
        return self._memoizedcall('_collected', lambda: list(self.collect()))
    ../lib/python3.6/site-packages/_pytest/python.py:412: in collect
        return super(Module, self).collect()
    ../lib/python3.6/site-packages/_pytest/python.py:332: in collect
        l.sort(key=lambda item: item.reportinfo()[:2])
    ../lib/python3.6/site-packages/_pytest/python.py:332: in <lambda>
        l.sort(key=lambda item: item.reportinfo()[:2])
    ../lib/python3.6/site-packages/_pytest/python.py:264: in reportinfo
        modpath = self.getmodpath()
    ../lib/python3.6/site-packages/_pytest/python.py:238: in getmodpath
        assert name.endswith(".py")
    E   AssertionError

--- a/AUTHORS
+++ b/AUTHORS
@@ -85,6 +85,7 @@ Justyna Janczyszyn
 Kale Kundert
 Katarzyna Jachim
 Kevin Cox
+Kodi B. Arfer
 Lee Kamentsky
 Lev Maximov
 Loic Esteve

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -12,6 +12,9 @@
 * Added documentation related to issue (`#1937`_)
   Thanks `@skylarjhdownes`_ for the PR.
 
+* Allow collecting files with any file extension as Python modules (`#2369`_).
+  Thanks `@Kodiologist`_ for the PR.
+
 *
 
 *
@@ -23,11 +26,13 @@
 .. _@skylarjhdownes: https://github.com/skylarjhdownes
 .. _@fabioz: https://github.com/fabioz
 .. _@metasyn: https://github.com/metasyn
+.. _@Kodiologist: https://github.com/Kodiologist
 
 
 .. _#1937: https://github.com/pytest-dev/pytest/issues/1937
 .. _#2276: https://github.com/pytest-dev/pytest/issues/2276
 .. _#2336: https://github.com/pytest-dev/pytest/issues/2336
+.. _#2369: https://github.com/pytest-dev/pytest/issues/2369
 
 
 3.0.7 (2017-03-14)

--- a/_pytest/python.py
+++ b/_pytest/python.py
@@ -3,6 +3,7 @@
 import fnmatch
 import inspect
 import sys
+import os
 import collections
 import math
 from itertools import count
@@ -235,8 +236,7 @@ class PyobjMixin(PyobjContext):
                 continue
             name = node.name
             if isinstance(node, Module):
-                assert name.endswith(".py")
-                name = name[:-3]
+                name = os.path.splitext(name)[0]
                 if stopatmodule:
                     if includemodule:
                         parts.append(name)

--- a/testing/python/collect.py
+++ b/testing/python/collect.py
@@ -841,6 +841,34 @@ class TestConftestCustomization:
         l = modcol.collect()
         assert '_hello' not in l
 
+    def test_issue2369_collect_module_fileext(self, testdir):
+        """Ensure we can collect files with weird file extensions as Python
+        modules (#2369)"""
+        # We'll implement a little finder and loader to import files containing
+        # Python source code whose file extension is ".narf".
+        testdir.makeconftest("""
+            import sys, os, imp
+            from _pytest.python import Module
+
+            class Loader:
+                def load_module(self, name):
+                    return imp.load_source(name, name + ".narf")
+            class Finder:
+                def find_module(self, name, path=None):
+                    if os.path.exists(name + ".narf"):
+                        return Loader()
+            sys.meta_path.append(Finder())
+
+            def pytest_collect_file(path, parent):
+                if path.ext == ".narf":
+                    return Module(path, parent)""")
+        testdir.makefile(".narf", """
+            def test_something():
+                assert 1 + 1 == 2""")
+        # Use runpytest_subprocess, since we're futzing with sys.meta_path.
+        result = testdir.runpytest_subprocess()
+        result.stdout.fnmatch_lines('*1 passed*')
+
 def test_setup_only_available_in_subdir(testdir):
     sub1 = testdir.mkpydir("sub1")
     sub2 = testdir.mkpydir("sub2")

